README.md
# weather

[![Build Status](https://app.travis-ci.com/lykmapipo/weather.svg?branch=master)](https://app.travis-ci.com/lykmapipo/weather)
[![Dependencies Status](https://david-dm.org/lykmapipo/weather.svg)](https://david-dm.org/lykmapipo/weather)
[![Coverage Status](https://coveralls.io/repos/github/lykmapipo/weather/badge.svg?branch=master)](https://coveralls.io/github/lykmapipo/weather?branch=master)
[![GitHub License](https://img.shields.io/github/license/lykmapipo/weather)](https://github.com/lykmapipo/weather/blob/develop/LICENSE)

[![Commitizen Friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Code Style](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)
[![npm version](https://img.shields.io/npm/v/@lykmapipo/weather)](https://www.npmjs.com/package/@lykmapipo/weather)

Helpers to fetch weather information from well know sources

## Requirements

- [NodeJS v14.5+](https://nodejs.org)
- [Npm v6.14+](https://www.npmjs.com/)

## Installation

```sh
npm install @lykmapipo/weather --save
```

## Usage

```js
import { fetchForecasts } from '@lykmapipo/weather';

const optns = { city: 'Dar Es Salaam' };
fetchForecasts(optns)
  .then(forecasts => { ... }) // => [{ weather: 'Light Rain', ... }, ... ]
  .catch(error => { ... });
```

## Forecast Schema

| Name               | Type      | Description                       |
| ------------------ | --------- | --------------------------------- |
| issuedAt           | `date`    | Forecast issued date and time     |
| date               | `date`    | Forecast date                     |
| weather            | `string`  | Weather description               |
| temperature        | `number`  | Temperature in Celsius            |
| minimumTemperature | `number`  | Minimum temperature in Celsius    |
| maximumTemperature | `number`  | Maximum temperature in Celsius    |
| relativeHumidity   | `number`  | Relative humidity                 |
| windDirection      | `string`  | Wind direstion                    |
| windSpeed          | `number`  | Wind speed in `km/h`              |
| sunRiseAt          | `date`    | Sun rise time                     |
| sunSetAt           | `date`    | Sun set time                      |
| moonRiseAt         | `date`    | Moon rise time                    |
| moonSetAt          | `date`    | Moon set time                     |
| present            | `boolean` | Flag whether its present forecast |

## Supported Sources

- [World Weather Information Service](https://worldweather.wmo.int/en/home.html)

## Testing

- Clone this repository

- Install all development dependencies

```sh
npm install
```

- Run example

```sh
npm run examples
```

- Then run test

```sh
npm test
```

## Contribute

It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## License

The MIT License (MIT)

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 0.3.8 (2021-09-06)

##### Chores

* **deps:**  force latest version & audit fix ([ed354745](https://github.com/lykmapipo/weather/commit/ed354745e9ac1e4605ea724668308ab370f5a74e))

##### Refactors

* **deps:**  use latest version ([79f88252](https://github.com/lykmapipo/weather/commit/79f882523541a9916f8d72daec59c449faeac1bc))

#### 0.3.7 (2021-02-10)

#### 0.3.6 (2021-01-06)

##### Chores

- **deps:** force latest version & audit fix ([80bb820b](https://github.com/lykmapipo/weather/commit/80bb820b118dd8b3d13856c1ea79c723e4a52b87))

#### 0.3.5 (2020-12-26)

##### Chores

- **deps:** force latest version & audit fix ([461eb865](https://github.com/lykmapipo/weather/commit/461eb865b450228c5a2079bc02b1b14cd7250bb5))

#### 0.3.4 (2020-12-17)

##### Chores

- **deps:** force latest version & audit fix ([2d28f40a](https://github.com/lykmapipo/weather/commit/2d28f40a5247eb7071c6875cbfb9436c46f0fdc1))

#### 0.3.3 (2020-12-06)

##### Chores

- **deps:** force latest version & audit fix ([5e902c99](https://github.com/lykmapipo/weather/commit/5e902c99b21cd3e08146733df72a31c12d354842))

#### 0.3.2 (2020-10-05)

##### Chores

- **deps:** force latest version & audit fix ([d0512ac0](https://github.com/lykmapipo/weather/commit/d0512ac05dbce223134661bb0a88a4edbed946ee))

#### 0.3.1 (2020-09-23)

#### 0.3.0 (2020-09-20)

##### Documentation Changes

- **readme:**
  - describe returned forecast schema ([632cb5e0](https://github.com/lykmapipo/weather/commit/632cb5e06a04a881fe347b55e165b3420d0327bd))
  - describe returned forecast schema ([a5f263ce](https://github.com/lykmapipo/weather/commit/a5f263ce1466ebfd279f4c1b837adefaf5277a0e))
  - update run example guide ([77a92fe7](https://github.com/lykmapipo/weather/commit/77a92fe7a2e9c5cf8720288b8b51eb699cd12bab))

##### New Features

- **utils:** add issued time on normalized forecast ([d0d588b4](https://github.com/lykmapipo/weather/commit/d0d588b47f9da767c09c85301eea01a1e52c0cfa))

##### Refactors

- **utils:** improve forecast schema ([ae7362e5](https://github.com/lykmapipo/weather/commit/ae7362e5bbffcc6a18c3b0c6f39e6c44bc9fc3a5))

##### Code Style Changes

- improve jsdocs ([d301ec63](https://github.com/lykmapipo/weather/commit/d301ec63c7d668e962bf5904be56ceef8aa1b37e))
- improve jsdocs ([a4aa94ef](https://github.com/lykmapipo/weather/commit/a4aa94efc62b467a8a14f39c0024633eceb52127))

#### 0.2.0 (2020-09-17)

##### Chores

- **deps:** use latest version ([07e0579d](https://github.com/lykmapipo/weather/commit/07e0579d0f9ece19717471d6c7492cafa2077fcd))

##### Bug Fixes

- bad http request ([7bb7b683](https://github.com/lykmapipo/weather/commit/7bb7b683c2fa8532cd3f7e30d0db1837cb604c7e))

#### 0.1.1 (2020-09-10)

##### Chores

- **deps:** force latest version & audit fix ([1c205cb9](https://github.com/lykmapipo/weather/commit/1c205cb97c57979aa4381897097d42c9d84fb528))

#### 0.1.0 (2020-09-09)

##### Chores

- **deps:**
  - recorect used pkg versions ([c50431f7](https://github.com/lykmapipo/weather/commit/c50431f73ec71c47f8d51c78c2910845b7cd0f65))
  - sync packages info ([c8860ac3](https://github.com/lykmapipo/weather/commit/c8860ac3ff9abc27079f0987b95cf6b2f03faa88))

##### Documentation Changes

- **readme:** update usage docs ([2fa99ec3](https://github.com/lykmapipo/weather/commit/2fa99ec32dc25f17eec536d73769221a05f326be))

##### New Features

- fetch present & week forecasts ([bea901e7](https://github.com/lykmapipo/weather/commit/bea901e7af02cd6008078d293edd95e45ff40c04))
- add cities list ([2b7a7f03](https://github.com/lykmapipo/weather/commit/2b7a7f03299acf356f0af49f160dbd23fc7f7424))
- setup & initialize ([b1594fc0](https://github.com/lykmapipo/weather/commit/b1594fc08055e8c25d697c0e79a15f9cd7fde9ef))
- **utils:**
  - normalize week forecasts ([cb00190c](https://github.com/lykmapipo/weather/commit/cb00190c68d3b2e2bb67b0584ac2f2a071a7c3ec))
  - normalize present weather ([5ac40332](https://github.com/lykmapipo/weather/commit/5ac40332ac9a31cbe754011294a9c62f68f540b2))
  - generate wwis link for a city ([bde42e33](https://github.com/lykmapipo/weather/commit/bde42e33f1a44b034bc9d4d6c13288915d8276c9))
  - lookup for city by well known name and id ([6068451f](https://github.com/lykmapipo/weather/commit/6068451fb1974d0c6bb0a1a58c56b75037f5faf2))
  - lookup for city details ([f91d409e](https://github.com/lykmapipo/weather/commit/f91d409e6f59b9df0a6e661c4278710328b56c3a))
- **forecast:**
  - fetch week forecasts ([64f64179](https://github.com/lykmapipo/weather/commit/64f641795d2c0bf96cde4d235574cd1e1d116ccb))
  - fetch present forecasts ([8f3c2b6d](https://github.com/lykmapipo/weather/commit/8f3c2b6de38c81bccdb14597bf102e8b26475518))

##### Refactors

- handle unknown city ([11d6a05e](https://github.com/lykmapipo/weather/commit/11d6a05e8d322d39524ead3593bb4ca1f05f0db7))
- normalize fetched present forecasts ([08306650](https://github.com/lykmapipo/weather/commit/083066505c0f6d53ddbebf878b972a73bacd3fd9))
- normalize fetched present forecast ([026dca1e](https://github.com/lykmapipo/weather/commit/026dca1e2a9eae36ec51f3cc0a471593d255b585))
- **utils:** normalize well known cities data ([30725c48](https://github.com/lykmapipo/weather/commit/30725c48b23bc7fc9cdd1bd850e20e1c1909b858))


test/index.spec.js
import { createReadStream } from 'fs';
import { expect, nock } from '@lykmapipo/test-helpers';

import {
  fetchPresentForecast,
  fetchWeekForecasts,
  fetchForecasts,
} from '../src';

const BASE_URL = 'https://worldweather.wmo.int/en/json';

describe('weather', () => {
  beforeEach(() => {
    nock.cleanAll();
  });

  it('should fetch present forecast', (done) => {
    nock(BASE_URL)
      .get('/present.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/present.json`);
      });

    fetchPresentForecast({ city: 'Dar Es Salaam' })
      .then((forecast) => {
        expect(forecast).to.exist.and.be.an('object');
        // expect(forecast.country).to.exist.and.be.a('string');
        expect(forecast.city).to.exist.and.be.a('string');
        expect(forecast.cityId).to.exist.and.be.a('number');
        expect(forecast.date).to.exist.and.be.a('date');
        expect(forecast.issuedAt).to.exist.and.be.a('date');
        expect(forecast.weather).to.exist.and.be.a('string');
        expect(forecast.temperature).to.exist.and.be.a('number');
        expect(forecast.relativeHumidity).to.exist.and.be.a('number');
        // expect(forecast.pressure).to.exist.and.be.a('number');
        expect(forecast.windDirection).to.exist.and.be.a('string');
        expect(forecast.windSpeed).to.exist.and.be.a('number');
        expect(forecast.sunRiseAt).to.exist.and.be.a('date');
        expect(forecast.sunSetAt).to.exist.and.be.a('date');
        // expect(forecast.moonRiseAt).to.exist.and.be.a('date');
        // expect(forecast.moonSetAt).to.exist.and.be.a('date');
        expect(forecast.present).to.be.true;
        done(null, forecast);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  it('should handle error when fetch present forecast', (done) => {
    nock(BASE_URL)
      .get('/present.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/present.json`);
      });

    fetchPresentForecast({ city: 'Dar' }).catch((error) => {
      expect(error).to.exist;
      expect(error.message).to.be.equal('Unknown City');
      done();
    });
  });

  it('should fetch week forecasts', (done) => {
    nock(BASE_URL)
      .get('/252_en.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/252_en.json`);
      });

    fetchWeekForecasts({ city: 'Dar Es Salaam' })
      .then((forecasts) => {
        expect(forecasts).to.exist.and.be.an('array');
        expect(forecasts[0]).to.exist.and.be.an('object');
        // expect(forecasts[0].country).to.exist.and.be.a('string');
        expect(forecasts[0].city).to.exist.and.be.a('string');
        expect(forecasts[0].cityId).to.exist.and.be.a('number');
        expect(forecasts[0].date).to.exist.and.be.a('date');
        expect(forecasts[0].issuedAt).to.exist.and.be.a('date');
        expect(forecasts[0].weather).to.exist.and.be.a('string');
        // expect(forecasts[0].temperature).to.exist.and.be.a('number');
        // expect(forecasts[0].relativeHumidity).to.exist.and.be.a('number');
        // expect(forecasts[0].pressure).to.exist.and.be.a('number');
        // expect(forecasts[0].windDirection).to.exist.and.be.a('string');
        // expect(forecasts[0].windSpeed).to.exist.and.be.a('number');
        // expect(forecasts[0].sunRiseAt).to.exist.and.be.a('date');
        // expect(forecasts[0].sunSetAt).to.exist.and.be.a('date');
        // expect(forecasts[0].moonRiseAt).to.exist.and.be.a('date');
        // expect(forecasts[0].moonSetAt).to.exist.and.be.a('date');
        expect(forecasts[0].present).to.be.false;
        done(null, forecasts);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  it('should handle error when fetch week forecasts', (done) => {
    nock(BASE_URL)
      .get('/252_en.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/252_en.json`);
      });

    fetchWeekForecasts({ city: 'Dar' }).catch((error) => {
      expect(error).to.exist;
      expect(error.message).to.be.equal('Unknown City');
      done();
    });
  });

  it('should handle error when fetch forecasts', (done) => {
    nock(BASE_URL)
      .get('/present.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/present.json`);
      });

    nock(BASE_URL)
      .get('/252_en.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/252_en.json`);
      });

    fetchForecasts({ city: 'Dar Es Salaam' })
      .then((forecasts) => {
        expect(forecasts).to.exist.and.be.an('array');
        expect(forecasts[0]).to.exist.and.be.an('object');
        // expect(forecasts[0].country).to.exist.and.be.a('string');
        expect(forecasts[0].city).to.exist.and.be.a('string');
        expect(forecasts[0].cityId).to.exist.and.be.a('number');
        expect(forecasts[0].date).to.exist.and.be.a('date');
        expect(forecasts[0].issuedAt).to.exist.and.be.a('date');
        expect(forecasts[0].weather).to.exist.and.be.a('string');
        // expect(forecasts[0].temperature).to.exist.and.be.a('number');
        // expect(forecasts[0].relativeHumidity).to.exist.and.be.a('number');
        // expect(forecasts[0].pressure).to.exist.and.be.a('number');
        // expect(forecasts[0].windDirection).to.exist.and.be.a('string');
        // expect(forecasts[0].windSpeed).to.exist.and.be.a('number');
        // expect(forecasts[0].sunRiseAt).to.exist.and.be.a('date');
        // expect(forecasts[0].sunSetAt).to.exist.and.be.a('date');
        // expect(forecasts[0].moonRiseAt).to.exist.and.be.a('date');
        // expect(forecasts[0].moonSetAt).to.exist.and.be.a('date');
        expect(forecasts[0].present).to.be.true;
        done(null, forecasts);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  it('should fetch forecasts', (done) => {
    nock(BASE_URL)
      .get('/present.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/present.json`);
      });

    nock(BASE_URL)
      .get('/252_en.json')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/252_en.json`);
      });

    fetchForecasts({ city: 'Dar' }).catch((error) => {
      expect(error).to.exist;
      expect(error.message).to.be.equal('Unknown City');
      done();
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });
});


test/utils.spec.js
import { mergeObjects } from '@lykmapipo/common';
import { expect } from '@lykmapipo/test-helpers';

import {
  DEFAULT_REQUEST_HEADERS,
  findCity,
  parseTime,
  normalizePresentForecast,
  normalizeWeekForecasts,
} from '../src/utils';

import presentCity from './fixtures/city.json';
import presentForecast from './fixtures/present.json';
import weekForecasts from './fixtures/252_en.json';

describe('utils', () => {
  it('should provide default request headers', () => {
    expect(DEFAULT_REQUEST_HEADERS).to.be.eql({
      accept: 'application/json',
      'content-type': 'application/json',
      'user-agent':
        'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36',
    });
  });

  it('should lookup cities information', () => {
    expect(findCity()).to.be.undefined;
    expect(findCity({})).to.be.undefined;
    expect(findCity({ name: '' })).to.be.undefined;
    expect(findCity({ name: null })).to.be.undefined;
    expect(findCity({ name: undefined })).to.be.undefined;
    expect(findCity({ name: 'Unknown' })).to.be.undefined;
    expect(findCity({ name: 'Dar Es Salaam' })).to.be.exist.and.be.an('object');
    expect(findCity({ name: '252' })).to.be.exist.and.be.an('object');
    expect(findCity({ name: 252 })).to.be.exist.and.be.an('object');
    expect(findCity({ name: 'Dar Es Salaam' })).to.be.eql(
      findCity({ name: '252' })
    );
    expect(findCity({ name: 'Dar Es Salaam' })).to.be.eql(
      findCity({ name: 252 })
    );
  });

  it('should normalize rise and set time', () => {
    let normalizeDate;

    normalizeDate = parseTime(undefined);
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(undefined, undefined);
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(new Date('2020-09-08T00:00:00.000Z'), undefined);
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(undefined, '06:20');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(undefined, '06');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(undefined, '06:');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(undefined, '20');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(undefined, ':20');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(new Date('2020-09-08T00:00:00.000Z'), '06:');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(new Date('2020-09-08T00:00:00.000Z'), ':20');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime('2020-09-08T00:00:90.000Z', '06:20');
    expect(normalizeDate).to.be.undefined;

    normalizeDate = parseTime(new Date('2020-09-08T00:00:00.000Z'), '06:20');
    expect(normalizeDate).to.exist.and.be.a('date');
  });

  it('should normalize present city forecast', () => {
    const city = mergeObjects(presentCity);
    const forecast = mergeObjects(presentForecast);

    const normalizedForecast = normalizePresentForecast(forecast, city);
    expect(normalizedForecast).to.exist.and.be.an('object');
    // expect(normalizedForecast.country).to.exist.and.be.a('string');
    expect(normalizedForecast.city).to.exist.and.be.a('string');
    expect(normalizedForecast.cityId).to.exist.and.be.a('number');
    expect(normalizedForecast.date).to.exist.and.be.a('date');
    expect(normalizedForecast.issuedAt).to.exist.and.be.a('date');
    expect(normalizedForecast.weather).to.exist.and.be.a('string');
    expect(normalizedForecast.temperature).to.exist.and.be.a('number');
    expect(normalizedForecast.relativeHumidity).to.exist.and.be.a('number');
    // expect(normalizedForecast.pressure).to.exist.and.be.a('number');
    expect(normalizedForecast.windDirection).to.exist.and.be.a('string');
    expect(normalizedForecast.windSpeed).to.exist.and.be.a('number');
    expect(normalizedForecast.sunRiseAt).to.exist.and.be.a('date');
    expect(normalizedForecast.sunSetAt).to.exist.and.be.a('date');
    // expect(normalizedForecast.moonRiseAt).to.exist.and.be.a('date');
    // expect(normalizedForecast.moonSetAt).to.exist.and.be.a('date');
    expect(normalizedForecast.present).to.be.true;
  });

  it('should normalize week city forecasts', () => {
    const city = mergeObjects(presentCity);
    const forecasts = mergeObjects(weekForecasts);

    const normalizedForecasts = normalizeWeekForecasts(forecasts, city);
    expect(normalizedForecasts).to.exist.and.be.an('array');
    expect(normalizedForecasts[0]).to.exist.and.be.an('object');
    // expect(normalizedForecasts[0].country).to.exist.and.be.a('string');
    expect(normalizedForecasts[0].city).to.exist.and.be.a('string');
    expect(normalizedForecasts[0].cityId).to.exist.and.be.a('number');
    expect(normalizedForecasts[0].date).to.exist.and.be.a('date');
    expect(normalizedForecasts[0].issuedAt).to.exist.and.be.a('date');
    expect(normalizedForecasts[0].weather).to.exist.and.be.a('string');
    // expect(normalizedForecasts[0].temperature).to.exist.and.be.a('number');
    // expect(normalizedForecasts[0].relativeHumidity).to.exist.and.be.a('number');
    // expect(normalizedForecasts[0].pressure).to.exist.and.be.a('number');
    // expect(normalizedForecasts[0].windDirection).to.exist.and.be.a('string');
    // expect(normalizedForecasts[0].windSpeed).to.exist.and.be.a('number');
    // expect(normalizedForecasts[0].sunRiseAt).to.exist.and.be.a('date');
    // expect(normalizedForecasts[0].sunSetAt).to.exist.and.be.a('date');
    // expect(normalizedForecasts[0].moonRiseAt).to.exist.and.be.a('date');
    // expect(normalizedForecasts[0].moonSetAt).to.exist.and.be.a('date');
    expect(normalizedForecasts[0].present).to.be.false;
  });
});


src/utils.js
import { get as getValue, find, map, toNumber, trim, values } from 'lodash';
import moment from 'moment';
import {
  compact,
  isNotValue,
  mergeObjects,
  parseDate,
} from '@lykmapipo/common';

import KNOWN_CITIES from './cities.json';

/**
 * @constant CITIES
 * @name CITIES
 * @description Well known cities for weather forecasts
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @private
 * @ignore
 */
export const CITIES = map(KNOWN_CITIES, (knownCity) => {
  // obtain and normalize city data
  const [cityId, city] = knownCity.split(':');

  // compact and return city data
  return mergeObjects({ city, cityId: toNumber(cityId) });
});

/**
 * @constant DEFAULT_REQUEST_HEADERS
 * @name DEFAULT_REQUEST_HEADERS
 * @description Default http request headers
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @private
 * @ignore
 */
export const DEFAULT_REQUEST_HEADERS = {
  accept: 'application/json',
  'content-type': 'application/json',
  'user-agent':
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36',
};

/**
 * @function findCity
 * @name findCity
 * @description Find cities details
 * @param {object} optns Valid lookup options
 * @param {object} optns.name Valid city name
 * @returns {object|undefined} found city or undefined
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * findCity({ name : 'Dar Es Salaam' });
 * // => { Country: 'United Republic of Tanzania', CityId: '252', ... }
 */
export const findCity = (optns) => {
  // ensure options
  const { name } = mergeObjects(optns);

  // ignore
  if (isNotValue(name)) {
    return undefined;
  }

  // lookup given city from well known cities
  const cityName = trim(name);
  const foundCity = find(CITIES, ({ city, cityId }) => {
    return (
      String(city) === String(cityName) || String(cityId) === String(cityName)
    );
  });

  // return found city
  return foundCity;
};

/**
 * @function parseTime
 * @name parseTime
 * @description Parse and normalize time string to date
 * @param {Date} date Valid base date
 * @param {string} time Valid time string
 * @param {string} [separator=':'] Time string separator
 * @returns {Date | undefined} normalized date time in UTC
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.3.0
 * @version 0.2.0
 * @static
 * @public
 * @example
 *
 * const date = '2020-09-08T00:00:00.000Z';
 * const time = '06:20';
 * parseTime(date, time);
 * // => '2020-09-07T06:20:00.000Z'
 */
export const parseTime = (date, time, separator = ':') => {
  // ensure date & time
  if (isNotValue(date) || isNotValue(time)) {
    return undefined;
  }

  // create base date moment
  let baseMoment = moment.utc(date);
  if (isNotValue(baseMoment.valueOf())) {
    return undefined;
  }

  // obtain hour and minutes from time string
  const [hours, minutes] = time.split(separator);
  if (isNotValue(hours) || isNotValue(minutes)) {
    return undefined;
  }

  // set hours and minutes to base moment
  baseMoment = baseMoment.hours(hours).minutes(minutes);

  // derive date time
  const derivedDate = isNotValue(baseMoment.valueOf())
    ? undefined
    : baseMoment.toDate();

  // return derived date time
  return derivedDate;
};

/**
 * @function normalizePresentForecast
 * @name normalizePresentForecast
 * @description Normalize present city forecast
 * @param {object} forecast Valid present forecast
 * @param {object} city Valid city details
 * @returns {object} normalized present city forecast
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const forecast = { present: { '193': { cityId: 252, ... } } };
 * const city = { cityId: 252, ... };
 * normalizePresentForecast(forecast, city);
 * // => { city: 'Dar Es Salaam', weather: 'Light Rain', ... }
 */
export const normalizePresentForecast = (forecast, city) => {
  // collect found present forecast
  const { present } = mergeObjects(forecast);
  const presentForecasts = compact([].concat(values(present)));

  // ensure present city
  const presentCity = mergeObjects(city);

  // find city present forecast
  let presentForecast = find(presentForecasts, (cityForecast) => {
    return String(cityForecast.cityId) === String(presentCity.cityId);
  });

  // normalize & convert
  const date = parseDate(getValue(presentForecast, 'issue'), 'YYYYMMDD');
  presentForecast = mergeObjects(presentCity, {
    date,
    issuedAt: parseDate(getValue(presentForecast, 'issue'), 'YYYYMMDDHH'),
    weather: getValue(presentForecast, 'wxdesc'),
    temperature: toNumber(getValue(presentForecast, 'temp')), //= > °C
    relativeHumidity: toNumber(getValue(presentForecast, 'rh')),
    windDirection: getValue(presentForecast, 'wd'),
    windSpeed: (toNumber(getValue(presentForecast, 'ws')) * 18) / 5, //= > km/h
    sunRiseAt: parseTime(date, getValue(presentForecast, 'sunrise')),
    sunSetAt: parseTime(date, getValue(presentForecast, 'sunset')),
    moonRiseAt: parseTime(date, getValue(presentForecast, 'moonrise')),
    moonSetAt: parseTime(date, getValue(presentForecast, 'moonset')),
    present: true,
  });

  // return given city present forecast
  return presentForecast;
};

/**
 * @function normalizeWeekForecasts
 * @name normalizeWeekForecasts
 * @description Normalize present city forecast
 * @param {object} forecasts Valid week forecasts
 * @param {object} city Valid city details
 * @returns {object[]} normalized week city forecast
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const forecast = { city: { forecast: { forecastDay: [ ... ] } } };
 * const city = { cityId: 252, ... };
 * normalizeWeekForecasts(forecast, city);
 * // => [{ city: 'Dar Es Salaam', weather: 'Light Rain', ... }, ... ]
 */
export const normalizeWeekForecasts = (forecasts, city) => {
  // collect found week forecasts
  const {
    city: {
      forecast: { issueDate, forecastDay },
    },
  } = mergeObjects({ forecast: { forecastDay: [] } }, forecasts);
  const weekForecasts = compact([].concat(forecastDay));

  // ensure present city
  const presentCity = mergeObjects(city);

  // normalize & convert
  const cityWeekForecasts = map(weekForecasts, (forecast) => {
    const date = parseDate(getValue(forecast, 'forecastDate'), 'YYYY-MM-DD');
    return mergeObjects(presentCity, {
      date,
      issuedAt: parseDate(issueDate, 'YYYY-MM-DD HH:mm:ss'),
      weather: getValue(forecast, 'weather'),
      temperature: toNumber(getValue(forecast, 'temp')), //= > °C
      minimumTemperature: toNumber(getValue(forecast, 'minTemp')), //= > °C
      maximumTemperature: toNumber(getValue(forecast, 'maxTemp')), //= > °C
      relativeHumidity: toNumber(getValue(forecast, 'rh')),
      windDirection: getValue(forecast, 'wd'),
      windSpeed: (toNumber(getValue(forecast, 'ws')) * 18) / 5, //= > km/h
      sunRiseAt: parseTime(date, getValue(forecast, 'sunrise')),
      sunSetAt: parseTime(date, getValue(forecast, 'sunset')),
      moonRiseAt: parseTime(date, getValue(forecast, 'moonrise')),
      moonSetAt: parseTime(date, getValue(forecast, 'moonset')),
      present: false,
    });
  });

  // return given city week forecast
  return cityWeekForecasts;
};
